software/contrib/preset_manager.py

```diff
@@ -0,0 +1,191 @@
+            self.screen = 0
+            self.update_ui()
+            
+            
+    def enter_button(self):
+        if self.screen < 1:
+            self.screen += 1
+        else:
+            None #Furthest forward screen already reached
+            
+        if self.screen == 1:
+            self.load_new_preset()
+            
+        self.save_state()
+
+
+    def update_current_value(self, values):
+        old_selected_preset = self.selected_preset
+        self.selected_preset = k1.read_position(7) - 1
+        if self.selected_preset != old_selected_preset: #When selected preset is changed, take a snapshot of k2
+            self.old_k2 = k2.read_position(self.resolution)
+        else:
+            if k2.read_position(self.resolution) != self.old_k2: #k2 must change to allow value to be updated to prevent updating every value while scrolling
+                values[self.selected_preset] = k2.read_position(self.resolution)
+                
+
+    def draw_current_preset(self):
+        oled.fill(0)
+        
+        name = self.presets[self.loaded_preset][0]
+        values = self.presets[self.loaded_preset][1]
+        
+        oled.text(name, int((OLED_WIDTH - (len(name) * CHAR_WIDTH)) / 2), 0, 1)
+        
+        self.update_current_value(values)
+        
+        for index, value in enumerate(values):
+            x = int((index % 3)*self.box_width)
+            y = int((index // 3)*self.box_height) + self.box_height - 1
+            oled.rect(x, y, self.box_width, self.box_height, 1) if index != self.selected_preset else oled.fill_rect(x, y, self.box_width, self.box_height, 1)
+            oled.text(str(value), (x + 2), (y + 2), 1 if index != self.selected_preset else 0)
+        oled.show()    
+        
+
+    def update_outputs(self):
+        values = self.presets[self.loaded_preset][1]
+        for cv_value in zip(cvs, values):
+            cv_value[0].voltage((cv_value[1] * self.output_multiplier))
+
+
+    def draw_preset_list(self):
+        oled.fill(0)
+        
+        positions = len(self.presets) + 1
+        self.hovered_preset = k1.read_position(positions)
+        
+        vertical_offset = -(self.hovered_preset * self.list_line_height) + 1
+        oled.fill_rect(0, 0, OLED_WIDTH, self.list_line_height, 1)
+        for index, preset in enumerate(self.presets):
+            name = preset[0]
+            oled.text(name, 0, vertical_offset, 1 if self.hovered_preset != index else 0)
+            
+            vertical_offset += self.list_line_height
+        if self.hovered_preset == (positions - 1):
+            colour = 0
+            self.create_new_preset = True
+        else:
+            colour = 1
+            self.create_new_preset = False
+        oled.text('NEW PRESET', 0, vertical_offset, colour)
+        oled.show()
+    
+
+    def load_new_preset(self):
+        if self.create_new_preset:
+            self.presets.append(self.blank_preset)
```

Comment:
If you create more than one preset during the existence of a class instance, aren't you going to end up with those presets sharing a reference to the same list of `[preset, [values]]`?

Reply:
If I understand what you mean, then no, because the presets are simply indexed in the list of presets rather than accessed in a dictionary structure, two presets could have the same name and/or values with no issues